sdk/typescript/introspector/scanner/typescript_module/ast.ts

```diff
@@ -0,0 +1,368 @@
+      throw new IntrospectionError(
+        `decorator at ${AST.getNodePosition(node)} should be a call expression, please use ${daggerDecorator}() instead.`,
+      )
+    }
+
+    return true
+  }
+
+  public getDecoratorArgument<T>(
+    node: ts.HasDecorators,
+    daggerDecorator: DaggerDecorators,
+    type: "string" | "object",
+    position = 0,
+  ): T | undefined {
+    const decorators = ts.getDecorators(node)
+    if (!decorators) {
+      return undefined
+    }
+
+    const decorator = decorators.find((d) =>
+      d.expression.getText().startsWith(daggerDecorator),
+    )
+    if (!decorator) {
+      return undefined
+    }
+
+    const argument = (decorator.expression as ts.CallExpression).arguments[
+      position
+    ]
+    if (!argument) {
+      return undefined
+    }
+
+    switch (type) {
+      case "string":
+        return argument.getText() as T
+      case "object":
+        return eval(`(${argument.getText()})`)
+    }
+  }
+
+  public unwrapTypeStringFromPromise(type: string): string {
+    if (type.startsWith("Promise<")) {
+      return type.slice("Promise<".length, -">".length)
+    }
+
+    if (type.startsWith("Awaited<")) {
+      return type.slice("Awaited<".length, -">".length)
+    }
+
+    return type
+  }
+
+  public unwrapTypeStringFromArray(type: string): string {
+    if (type.endsWith("[]")) {
+      return type.replace("[]", "")
+    }
+
+    if (type.startsWith("Array<")) {
+      return type.slice("Array<".length, -">".length)
+    }
+
+    return type
+  }
+
+  public stringTypeToUnwrappedType(type: string): string {
+    type = this.unwrapTypeStringFromPromise(type)
+
+    // If they are difference, that means we upwrapped the array.
+    const extractedTypeFromArray = this.unwrapTypeStringFromArray(type)
+    if (extractedTypeFromArray !== type) {
+      return this.stringTypeToUnwrappedType(extractedTypeFromArray)
+    }
+
+    return type
+  }
+
+  public typeToStringType(type: ts.Type): string {
+    const stringType = this.checker.typeToString(type)
+
+    return this.stringTypeToUnwrappedType(stringType)
+  }
+
+  public tsTypeToTypeDef(
+    node: ts.Node,
+    type: ts.Type,
+  ): TypeDef<TypeDefKind> | undefined {
+    if (type.flags & ts.TypeFlags.String)
+      return { kind: TypeDefKind.StringKind }
+    if (type.flags & ts.TypeFlags.Number)
+      return { kind: TypeDefKind.IntegerKind }
+    if (type.flags & ts.TypeFlags.Boolean)
+      return { kind: TypeDefKind.BooleanKind }
+    if (type.flags & ts.TypeFlags.Void) return { kind: TypeDefKind.VoidKind }
+
+    // If a type has a flag Object, is can basically be anything.
+    // We firstly wants to see if it's a promise or an array so we can unwrap the
+    // actual type.
+    if (type.flags & ts.TypeFlags.Object) {
+      const objectType = type as ts.ObjectType
+
+      // If it's a reference, that means it's a generic type like
+      // `Promise<T>` or `number[]` or `Array<T>`.
```

Comment:
These explanations, especially when showing what you see in code like in the second line, and especially when it's not something obvious like a string, are very useful. Please do more of these 👍